### PR TITLE
CAS-171 Excessive errors take a child out of the IO path

### PR DESCRIPTION
**Pull request by chriswldenyer**
When an IO is added to error store, the error store is queried
to determine whether the number of errors meets the criterion that
should cause the child to be marked as faulted. If so, the child is
faulted.
The error store can also determine the number of repeated attempts
to send the same IO that have failed. This enables a per-IO error
count to be queried.
Configuration options exist to enable the error store and to enable
faulting the child due to number of retries or total number of IO
errors, and dependent on the age of the error.
Existing unit test to verify new behaviour and added unit tests for
child removal due to total error count and number of repeated errors.

### CODE REVIEW

File: mayastor/tests/error_fault_child_total.rs [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+    common::delete_file(&[DISKNAME2.to_string()]);
+    common::delete_file(&[YAML_CONFIG_FILE.to_string()]);
+}
+
+fn check_nexus_state_is(expected_status: NexusStatus) {
+    let nexus = nexus_lookup(ERROR_COUNT_TEST_NEXUS).unwrap();
+    assert_eq!(nexus.status(), expected_status);
+}
+
+async fn create_error_bdev() {
```
Comment (jonathan-teh) [inline mayastor/tests/error_fault_child_total.rs:103]: There is a common version of this in common/error_bdev.rs.

File: mayastor/tests/error_fault_child_total.rs [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+    let err_bdev_name_str =
+        CString::new(EE_ERROR_DEVICE).expect("Failed to create name string");
+    let raw = err_bdev_name_str.into_raw();
+
+    unsafe {
+        retval = spdk_vbdev_error_inject_error(raw, op, mode, count);
+    }
+    assert_eq!(retval, 0);
+}
+fn reactor_run_millis(milliseconds: u64) {
```
Comment (jonathan-teh) [inline mayastor/tests/error_fault_child_total.rs:185]: Nit: missing blank line.

File: mayastor/tests/error_fault_child_retries.rs [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+    let err_bdev_name_str =
+        CString::new(EE_ERROR_DEVICE).expect("Failed to create name string");
+    let raw = err_bdev_name_str.into_raw();
+
+    unsafe {
+        retval = spdk_vbdev_error_inject_error(raw, op, mode, count);
+    }
+    assert_eq!(retval, 0);
+}
+fn reactor_run_millis(milliseconds: u64) {
```
Comment (jonathan-teh) [inline mayastor/tests/error_fault_child_retries.rs:186]: Nit: blank line.

File: mayastor/src/bdev/nexus/nexus_child_error_store.rs [hunk 4]
```diff
@@ -48,6 +51,11 @@ pub struct NexusErrStore {
     records: Vec<NexusChildErrorRecord>,
 }
 
+pub enum NexusErrStoreQuery {
```
Comment (gila) [inline mayastor/src/bdev/nexus/nexus_child_error_store.rs:54]: as all the code is already namespaced (i.e bdev::nexus::nexus_child_error_store) there is no need to make that:

bdev::nexus::nexus_child_error_store::NexusErrStoreQuery -- that is all very, very verbose, would you agree? Ideally it would not have been named nexus_child_error_store either but thats not part of this change so thats OK.

File: mayastor/src/bdev/nexus/nexus_child_error_store.rs [hunk 5]
```diff
@@ -48,6 +51,11 @@ pub struct NexusErrStore {
...
+    Total,
```
Comment (gila) [inline mayastor/src/bdev/nexus/nexus_child_error_store.rs:55]: please document the members and the purpose

File: mayastor/src/bdev/nexus/nexus_child_error_store.rs [hunk 6]
```diff
@@ -105,61 +113,138 @@ impl NexusErrStore {
         self.next_record_index =
              // consider using a mask if we restrict the size to a power-of-2
             (self.next_record_index + 1) % self.records.len();
-        // debug!("added record - buffer is {}", &self);
     }
 
-    pub fn query(
+    #[inline]
+    fn is_wanted_record(
```
Comment (gila) [inline mayastor/src/bdev/nexus/nexus_child_error_store.rs:119]: would "find" be better? or maybe, has_error(..) ?

File: mayastor/src/bdev/nexus/nexus_child_error_store.rs [hunk 7]
```diff
@@ -105,61 +113,138 @@ impl NexusErrStore {
...
+        record: &NexusChildErrorRecord,
+        io_op_flags: u32,
+        io_error_flags: u32,
+        target_timestamp: Option<Instant>,
+    ) -> bool {
+        let found_op = match record.io_op {
+            io_type::READ => (io_op_flags & NexusErrStore::READ_FLAG) != 0,
```
Comment (gila) [inline mayastor/src/bdev/nexus/nexus_child_error_store.rs:126]: this can be refined using match guards: https://doc.rust-lang.org/reference/expressions/match-expr.html#match-guards

File: mayastor/src/bdev/nexus/nexus_child_error_store.rs [hunk 8]
```diff
@@ -105,61 +113,138 @@ impl NexusErrStore {
...
+            io_type::WRITE => (io_op_flags & NexusErrStore::WRITE_FLAG) != 0,
+            io_type::UNMAP => (io_op_flags & NexusErrStore::UNMAP_FLAG) != 0,
+            io_type::FLUSH => (io_op_flags & NexusErrStore::FLUSH_FLAG) != 0,
+            io_type::RESET => (io_op_flags & NexusErrStore::RESET_FLAG) != 0,
+            _ => false,
```
Comment (gila) [inline mayastor/src/bdev/nexus/nexus_child_error_store.rs:131]: if you return here, the if condition on line 134 is not needed

File: mayastor/src/bdev/nexus/nexus_child_error_store.rs [hunk 9]
```diff
@@ -105,61 +113,138 @@ impl NexusErrStore {
...
+        };
+
+        if found_op {
+            let found_err = match record.io_error {
+                io_status::FAILED => {
+                    (io_error_flags & NexusErrStore::IO_FAILED_FLAG) != 0
+                }
+                _ => false,
```
Comment (gila) [inline mayastor/src/bdev/nexus/nexus_child_error_store.rs:139]: the same here, an early return would make the code a bit cleaner?

File: mayastor/src/bdev/nexus/nexus_child_error_store.rs [hunk 10]
```diff
@@ -105,61 +113,138 @@ impl NexusErrStore {
...
-                _ => false,
-            };
+        }
+        err_count
+    }
 
-            let found_err = match self.records[idx].io_error {
-                io_status::FAILED => {
-                    (io_error_flags & NexusErrStore::IO_FAILED_FLAG) != 0
+    #[inline]
```
Comment (gila) [inline mayastor/src/bdev/nexus/nexus_child_error_store.rs:181]: I'm not quite sure what we are trying to do here...  If we want to count the errors for a particular type, this seems rather expensive.

File: mayastor/src/subsys/opts.rs [hunk 11]
```diff
@@ -482,15 +482,36 @@ impl GetOpts for IscsiTgtOpts {
 pub struct ErrStoreOpts {
     /// ring buffer size
     pub err_store_size: usize,
+
     /// NexusErrStore enabled
     pub enable_err_store: bool,
+
+    /// whether to fault the child due to failed re-tried IOs
+    pub fault_child_on_retry_count: bool,
```
Comment (gila) [inline mayastor/src/subsys/opts.rs:490]: I think it would be better to name this a particular action we can do, i.e retries_exceed = "action" where action is ingore, reset, fault, or whatever.

File: mayastor/src/subsys/opts.rs [hunk 12]
```diff
@@ -482,15 +482,36 @@ impl GetOpts for IscsiTgtOpts {
...
+
+    /// the maximum number of errors for the same retried IO
+    pub max_retry_errors: u32,
```
Comment (gila) [inline mayastor/src/subsys/opts.rs:493]: max_retries?

File: mayastor/src/subsys/opts.rs [hunk 13]
```diff
@@ -482,15 +482,36 @@ impl GetOpts for IscsiTgtOpts {
...
+
+    /// whether to fault the child due to the total number of failed IOs
+    pub fault_child_on_error_count: bool,
```
Comment (gila) [inline mayastor/src/subsys/opts.rs:496]: I'm not sure we want to make our lives so hard here.. Whenever an IO failed WITH retires or not, it a failed IO that has been retried. 

Put differently, an IO to LBA0 that fails, retried, fails, retried OK, is still 2 IO errors. right? The Errorstore contains more detail as to what why how *exactly* but that's more for diagnosing the core problem?

To keep it simple, I would suggest to start out simple, getting the right behavior on failed IO's to a child, would be WIN-WIN, then we can start to think about special cases

File: mayastor/src/subsys/opts.rs [hunk 14]
```diff
@@ -482,15 +482,36 @@ impl GetOpts for IscsiTgtOpts {
...
+
+    /// the maximum number of errors in total
+    pub max_total_errors: u32,
+
+    /// errors older than this are ignored
+    pub max_error_age_ns: u64,
```
Comment (gila) [inline mayastor/src/subsys/opts.rs:502]: retention?

File: mayastor/src/subsys/opts.rs [hunk 15]
```diff
@@ -482,15 +482,36 @@ impl GetOpts for IscsiTgtOpts {
...
 
 impl Default for ErrStoreOpts {
     fn default() -> Self {
         Self {
             err_store_size: 256,
-            enable_err_store: true,
+            enable_err_store: false,
+            fault_child_on_retry_count: false,
+            fault_child_on_error_count: false,
+            max_retry_errors: 0,
```
Comment (gila) [inline mayastor/src/subsys/opts.rs:512]: is it a good idea to set these to 0? one retry and your out?

File: mayastor/tests/error_count.rs [hunk 16]
```diff
@@ -35,12 +36,19 @@ static ERROR_DEVICE: &str = "error_device";
...
+
 #[test]
 fn nexus_error_count_test() {
     common::truncate_file(DISKNAME1, 64 * 1024);
     common::truncate_file(DISKNAME2, 64 * 1024);
 
-    test_init!();
+    let mut config = Config::default();
+    config.err_store_opts.enable_err_store = true;
+    config.err_store_opts.err_store_size = 256;
```
Comment (gila) [inline mayastor/tests/error_count.rs:48]: if not part of this commit then please ignore, but "num_entries" would be a more descriptive name.

File: mayastor/tests/error_count.rs [hunk 17]
```diff
@@ -35,12 +36,19 @@ static ERROR_DEVICE: &str = "error_device";
...
+    config.write(YAML_CONFIG_FILE).unwrap();
+
+    test_init!(Some(YAML_CONFIG_FILE.to_string()));
```
Comment (gila) [inline mayastor/tests/error_count.rs:51]: Its not some to see Some(macro!()) perhaps, give the testing, you want to unwrap in the macro?

File: mayastor/src/subsys/opts.rs (hunk 13 above)
Comment (chriswldenyer) [reply mayastor/src/subsys/opts.rs:496]: My reasoning was that we would want to distinguish between, say, 10 different IOs which all fail on the first attempt then succeed, and one IO that fails 10 times in a row. It is easy to imagine wanting different thresholds for the two cases.
But I can take out the repeated IO error handling for now.

File: mayastor/src/bdev/nexus/nexus_child_error_store.rs (hunk 4 above)
Comment (chriswldenyer) [reply mayastor/src/bdev/nexus/nexus_child_error_store.rs:54]: I tend to prefer verbose names, but if the house style is to keep them short, I'll do so.

File: mayastor/tests/error_fault_child_retries.rs [hunk 18]
```diff
@@ -0,0 +1,193 @@
...
+    common::delete_file(&[DISKNAME2.to_string()]);
+    common::delete_file(&[YAML_CONFIG_FILE.to_string()]);
+}
+
+fn check_nexus_state_is(expected_status: NexusStatus) {
+    let nexus = nexus_lookup(ERROR_COUNT_TEST_NEXUS).unwrap();
+    assert_eq!(nexus.status(), expected_status);
+}
+
+async fn create_error_bdev() {
```
[the remark this answers is not in the document]
Comment (chriswldenyer) [reply mayastor/tests/error_fault_child_retries.rs:104]: No, Jonathan pointed that out as well. I am changing it to use the common function.

File: mayastor/tests/error_fault_child_retries.rs [hunk 19]
```diff
@@ -0,0 +1,193 @@
...
+        Ok(_) => {
+            assert_eq!(succeed, true);
+        }
+        Err(_) => {
+            assert_eq!(succeed, false);
+        }
+    };
+}
+
+async fn inject_error(op: u32, mode: u32, count: u32) {
```
[the remark this answers is not in the document]
Comment (chriswldenyer) [reply mayastor/tests/error_fault_child_retries.rs:175]: Also being fixed as above

File: mayastor/src/bdev/nexus/nexus_bdev_children.rs [hunk 20]
```diff
@@ -259,6 +260,31 @@ impl Nexus {
...
+            return Err(Error::DestroyLastChild {
+                name: self.name.clone(),
+                child: name.to_owned(),
+            });
+        }
+        if let Some(child) = self.children.iter_mut().find(|c| c.name == name) {
+            if child.status() != ChildStatus::Faulted {
+                child.fault();
+                self.reconfigure(DREvent::ChildFault).await;
+                info!("Fault and reconfigure completed");
```
[the remark this answers is not in the document]
Comment (chriswldenyer) [reply mayastor/src/bdev/nexus/nexus_bdev_children.rs:277]: Ok, will remove this.

File: mayastor/src/bdev/nexus/nexus_child_error_store.rs [hunk 21]
```diff
@@ -303,4 +424,22 @@ impl Nexus {
             Ok(None)
         }
     }
+
+    // returns false if the child is deemed faulted
```
[the remark this answers is not in the document]
Comment (chriswldenyer) [reply mayastor/src/bdev/nexus/nexus_child_error_store.rs:428]: Am expanding the comment.

File: mayastor/src/bdev/nexus/nexus_child_error_store.rs [hunk 22]
```diff
@@ -241,23 +327,56 @@ impl Nexus {
...
+                            if nexus.fault_child(&child_name).await.is_err() {
+                                error!(
+                                    "Failed to fault the child {}",
+                                    child_name,
+                                );
+                            }
+                        }
+                    } else {
+                        error!(
+                            "Failed to record error - child has no error store"
```
[the remark this answers is not in the document]
Comment (chriswldenyer) [reply mayastor/src/bdev/nexus/nexus_child_error_store.rs:373]: Added child name to the message.

File: mayastor/src/bdev/nexus/nexus_bdev_children.rs [hunk 23]
```diff
@@ -259,6 +260,31 @@ impl Nexus {
         Ok(self.status())
     }
 
+    /// fault a child device and reconfigure the IO channels
+    pub async fn fault_child(&mut self, name: &str) -> Result<(), Error> {
+        trace!("{}: fault child request for {}", self.name, name);
+
+        if self.child_count < 2 {
+            return Err(Error::DestroyLastChild {
```
[the remark this answers is not in the document]
Comment (chriswldenyer) [reply mayastor/src/bdev/nexus/nexus_bdev_children.rs:268]: No, but following discussion, will rename error and change error message.

File: mayastor/tests/error_fault_child_total.rs (hunk 1 above)
Comment (chriswldenyer) [reply mayastor/tests/error_fault_child_total.rs:103]: Now using the common version of this function (and inject_error()) in this and error_fault_child_retries.rs

File: mayastor/tests/error_fault_child_total.rs (hunk 2 above)
Comment (chriswldenyer) [reply mayastor/tests/error_fault_child_total.rs:185]: fixed

File: mayastor/tests/error_fault_child_retries.rs (hunk 3 above)
Comment (chriswldenyer) [reply mayastor/tests/error_fault_child_retries.rs:186]: fixed

File: mayastor/src/subsys/opts.rs (hunk 14 above)
Comment (chriswldenyer) [reply mayastor/src/subsys/opts.rs:502]: changed to retention_ns

File: mayastor/src/bdev/nexus/nexus_child_error_store.rs (hunk 5 above)
Comment (chriswldenyer) [reply mayastor/src/bdev/nexus/nexus_child_error_store.rs:55]: Enum purpose and values now described in comments.

File: mayastor/tests/error_fault_child_total.rs [hunk 24]
```diff
@@ -0,0 +1,192 @@
+extern crate log;
```
[the remark this answers is not in the document]
Comment (chriswldenyer) [reply mayastor/tests/error_fault_child_total.rs:1]: Removing retry-count functionality means only one file is needed, so this is no longer an issue (for now).

File: mayastor/src/subsys/opts.rs (hunk 15 above)
Comment (chriswldenyer) [reply mayastor/src/subsys/opts.rs:512]: I'd expect this to be set to some non-zero value if the feature is enabled but I didn't want to assign an arbitrary number. 
On the other hand, a common retry value in SPDK is 5, which seems reasonable, so I'll set that as the default.

File: mayastor/src/bdev/nexus/nexus_child_error_store.rs (hunk 7 above)
Comment (chriswldenyer) [reply mayastor/src/bdev/nexus/nexus_child_error_store.rs:126]: Refactored using match guards.

File: mayastor/src/bdev/nexus/nexus_child_error_store.rs (hunk 8 above)
Comment (chriswldenyer) [reply mayastor/src/bdev/nexus/nexus_child_error_store.rs:131]: Added early return.

File: mayastor/src/bdev/nexus/nexus_child_error_store.rs (hunk 9 above)
Comment (chriswldenyer) [reply mayastor/src/bdev/nexus/nexus_child_error_store.rs:139]: added early return

File: mayastor/src/bdev/nexus/nexus_child_error_store.rs (hunk 6 above)
Comment (chriswldenyer) [reply mayastor/src/bdev/nexus/nexus_child_error_store.rs:119]: Ok, I'll go with has_error().

File: mayastor/src/subsys/opts.rs (hunk 12 above)
Comment (chriswldenyer) [reply mayastor/src/subsys/opts.rs:493]: This option is now not needed and removed.
But in the same vein renamed the other option max_total_errors to max_errors.

File: mayastor/tests/error_count.rs (hunk 16 above)
Comment (chriswldenyer) [reply mayastor/tests/error_count.rs:48]: It wasn't part of this commit. But I'll consider renaming it in a future iteration.

File: mayastor/src/subsys/opts.rs (hunk 11 above)
Comment (chriswldenyer) [reply mayastor/src/subsys/opts.rs:490]: Changed variable from a bool to an enum with currently 2 values, "fault" and "ignore"

File: mayastor/tests/error_count.rs (hunk 17 above)
Comment (chriswldenyer) [reply mayastor/tests/error_count.rs:51]: Moved the string handling into the macro.

File: mayastor/src/bdev/nexus/nexus_child_error_store.rs (hunk 10 above)
Comment (chriswldenyer) [reply mayastor/src/bdev/nexus/nexus_child_error_store.rs:181]: count_repeats() now taken out of this PR as a result of taking out the repeated IO error handling, as suggested in this review.
